src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -278,6 +285,11 @@ private void WatchForFileSystemEventsThreadStart(object arg)
                 {
                     // Always unschedule the RunLoop before cleaning up
                     Interop.EventStream.FSEventStreamUnscheduleFromRunLoop(_eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+
+                    // Release the WatcherLoop Core Foundation object.
+                    CFStringRef release = _watcherRunLoop;
+                    _watcherRunLoop = IntPtr.Zero;
+                    Interop.CoreFoundation.CFRelease(release);
```

Comment:
Seems like this would need to be synchronized with cancellation.  Otherwise cancellation could read the value of _watcherRunLoop before it's set to zero, then this thread could release it, then the cancellation code could end up trying to use an already released value, and we could still crash.